cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ApplicationFilter.java

```diff
@@ -110,4 +134,29 @@ public String toString() {
         '}';
     }
   }
+
+  /**
+   * Returns true if the application version is latest
+   */
+  public static class ApplicationLatestFilter extends ApplicationIdFilter {
+    private final ApplicationLifecycleService applicationLifecycleService;
+
+    public ApplicationLatestFilter(ApplicationLifecycleService applicationLifecycleService) {
+      this.applicationLifecycleService = applicationLifecycleService;
+    }
+
+    @Override
+    public boolean test(ApplicationId applicationId) {
+      return applicationId.getVersion().equals(applicationLifecycleService.getLatestAppVersion(
```

Comment:
I don't think this would work fast enough for a lot of applications. We already had problems with a lot of apps in namespace. Doing a query for each line is pretty expensive

Reply:
yeah can we pass the latest version directly?

Reply:
All these filter are post query filter? That could result in retrieving a lot of data from DB? We should be able to use query to do it?

Reply:
During last set of optimization it was not pushed all the way to SQL (lack of time to do it). But this filters are applied before we deserialize JSON, that gave us a really good perf boost.

Reply:
fixed to use new scanByIndex method